tests/mgbench/client.cpp

```diff
@@ -226,12 +290,94 @@ void Execute(
   summary["retries"] = final_retries;
   summary["metadata"] = final_metadata.Export();
   summary["num_workers"] = FLAGS_num_workers;
+  summary["latency_stats"] = LatencyStatistics(worker_query_durations);
+  (*stream) << summary.dump() << std::endl;
+}
+
+nlohmann::json BoltRecordsToJSONStrings(std::vector<std::vector<memgraph::communication::bolt::Value>> &results) {
+  nlohmann::json res = nlohmann::json::object();
+  std::ostringstream oss;
+  for (int i = 0; i < results.size(); i++) {
+    oss << results[i];
+    res[std::to_string(i)] = oss.str();
+  }
+  return res;
+}
+
+// Validation mode works on single thread with 1 query.
+void Execute_validation(
+    const std::vector<std::pair<std::string, std::map<std::string, memgraph::communication::bolt::Value>>> &queries,
+    std::ostream *stream) {
+  spdlog::info("Running validation mode, number of workers forced to 1");
+  FLAGS_num_workers = 1;
+
+  std::thread thread;
+  Metadata worker_metadata = Metadata();
+  double worker_duration = 0.0;
+  std::vector<std::vector<memgraph::communication::bolt::Value>> worker_results;
+
+  auto size = queries.size();
+
+  thread = std::thread([&]() {
+    memgraph::io::network::Endpoint endpoint(FLAGS_address, FLAGS_port);
+    memgraph::communication::ClientContext context(FLAGS_use_ssl);
+    memgraph::communication::bolt::Client client(context);
+    client.Connect(endpoint, FLAGS_username, FLAGS_password);
+
+    auto &results = worker_results;
+    auto &metadata = worker_metadata;
+    auto &duration = worker_duration;
+
+    memgraph::utils::Timer timer;
+    if (size == 1) {
+      const auto &query = queries[0];
+      auto ret = ExecuteValidationNTimesTillSuccess(&client, query.first, query.second, FLAGS_max_retries);
+      metadata.Append(ret.first);
+      results = ret.second;
+      duration = timer.Elapsed().count();
+      client.Close();
+    } else {
+      spdlog::info("Validation works with single query, pass just one query!");
+    }
+  });
+
+  thread.join();
```

Comment:
All this can happen on the main thread, in other words you don't need an additional thread 🤔